adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+			if err != nil {
+				return nil, err
+			}
+			bid.ADomain = append(bid.ADomain, "huaweiads")
+			bid.NURL = getNurl(&content)
+			bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: mapSlotid2MediaType[ad30.Slotid],
+			})
+		}
+	}
+	return bidderResponse, nil
+}
+
+// get nurl when win auction.
+func getNurl(content *content) string {
+	if len(content.Monitor) == 0 {
+		return ""
+	}
+	for _, monitor := range content.Monitor {
+		if monitor.EventType == "win" && len(monitor.Url) != 0 {
+			return monitor.Url[0]
+		}
+	}
+	return ""
+}
+
+// get field Adm, Width, Height
+func (a *adapter) handleHuaweiAdsContent(adType int32, content *content, bidType openrtb_ext.BidType, imp openrtb2.Imp) (
+	adm string, adWidth int64, adHeight int64, err error) {
+	// v1: only support banner, native
+	switch bidType {
+	case openrtb_ext.BidTypeBanner:
+		adm, adWidth, adHeight, err = a.extractAdmBanner(adType, content, bidType, imp)
+	case openrtb_ext.BidTypeNative:
+		adm, adWidth, adHeight, err = a.extractAdmNative(adType, content, bidType, imp)
+	case openrtb_ext.BidTypeVideo:
+		adm, adWidth, adHeight, err = a.extractAdmVideo(adType, content, bidType, imp)
+	default:
+		return "", 0, 0, fmt.Errorf("no support bidtype: audio")
+	}
+
+	if err != nil {
+		return "", 0, 0, fmt.Errorf("getAdmFromHuaweiAdsContent failed: %s", err)
+	}
+	return adm, adWidth, adHeight, nil
+}
+
+// banner ad
+func (a *adapter) extractAdmBanner(adType int32, content *content, bidType openrtb_ext.BidType, imp openrtb2.Imp) (adm string,
+	adWidth int64, adHeight int64, err error) {
+	if adType != banner {
+		return "", 0, 0, fmt.Errorf("extractAdmBanner: huaweiads response is not a banner ad")
+	}
+	var creativeType = content.Creativetype
+	if content.Creativetype > 100 {
+		creativeType = creativeType - 100
+	}
+	if creativeType == text || creativeType == bigPicture || creativeType == bigPicture2 ||
+		creativeType == smallPicture || creativeType == threeSmallPicturesText ||
+		creativeType == iconText || creativeType == gif {
+		return a.extractAdmPicture(content)
+	} else if creativeType == videoText || creativeType == video || creativeType == videoWithPicturesText {
+		return a.extractAdmVideo(adType, content, bidType, imp)
+	} else {
+		return "", 0, 0, fmt.Errorf("no banner support creativetype")
+	}
+}
+
+// native ad
+func (a *adapter) extractAdmNative(adType int32, content *content, bidType openrtb_ext.BidType, openrtb2Imp openrtb2.Imp) (adm string,
+	adWidth int64, adHeight int64, err error) {
+	if adType != native {
+		return "", 0, 0, fmt.Errorf("extractAdmNative: response is not a native ad")
+	}
+	if openrtb2Imp.Native == nil {
+		return "", 0, 0, fmt.Errorf("extractAdmNative: imp.Native is nil")
+	}
+	if openrtb2Imp.Native.Request == "" {
+		return "", 0, 0, fmt.Errorf("extractAdmNative: imp.Native.Request is empty")
+	}
```

Comment:
Please add test coverage for these three if statements.

Reply:
added missing_native_request.json

Reply:
add bad_response_not_native.json